### PR TITLE
Show RNA and protein diagrams in related genes ideogram (SCP-4887)

### PR DESCRIPTION
**Pull request by eweitz**
This adds interactive RNA and protein diagrams to related genes ideogram.  It gives granular functional insight for genes.

### Overview 
Previously, tooltips shown upon hovering over a gene had only text: the gene's full name, genomic coordinates, and its relation to the searched gene.  Now, for protein-coding genes in human, diagrams are shown too.  Mature mRNA is at top, and aligned features in the corresponding protein are on bottom.  Hovering over parts of each diagram shows details, e.g. its length, and exon number or name.

#### Splice toggle
Clicking the scissors icon toggles the pre-mRNA into view.  While the _mature_ mRNA shown by default is basically equivalent to the cDNA molecule that is assayed in RNA-seq experiments, the pre-mRNA (i.e. preliminary mRNA) includes introns.

#### Alternative transcripts and isoforms
These diagrams exist for both canonical and alternative transcripts.  The [canonical transcript](https://useast.ensembl.org/info/genome/genebuild/canonical.html) is the default RNA transcript shown in traditional genome browsers.  Genes can also have other transcripts, e.g. due to alternative splicing or RNA editing.  These mRNA molecules and their corresponding proteins are often called "isoforms".  Different transcripts of the same gene are sometimes expressed in a cell type-specific manner.

### Use cases
These diagrams add value by giving granular functional insight and molecular references for genes.  For example:

* ***Easily learn more about what the gene does***.  Say you hover over a protein feature and see "GTP-binding domain".   Now you know the protein binds GTP!  Or maybe the protein feature notes it has kinase activity, and that's not deducible from the protein's full name.  Perhaps this helps you -- a biologist -- make other useful connections based on your scientific knowledge.  The new protein feature names convey key functional information about the gene you're looking at.

* ***Gauge known transcriptional complexity***.  Does your gene have 1 known transcript, or 10+?  Seeing many alternative transcripts in the "Transcripts" dropdown might warrant more care before assuming that your expression was due to the canonical transcript.

* ***Assess paralog similarity***.  Knowing that two genes are paralogs means knowing they are similar -- and now you can assess at a glance _how similar_ those genes are.  Does one gene have a much larger 3'-UTR?  Are their protein features basically the same, quite different, or somewhere in the middle?

There are potential future opportunities with sample data, e.g. integrating BAMs.  As is, though, these diagrams give unique and convenient biochemical context for genes.

Data on exons and transcripts comes from Ensembl, data on protein features comes from InterPro via Ensembl.

### <span id="video">Video</span>
Here's how it looks:

https://user-images.githubusercontent.com/1334561/212961840-1781d43f-9c1e-4bb3-95ec-2eb8c488aae7.mov

### Test
To manually verify:

1. **If local**, pull this branch, boot SCP via `bin/docker-compose-setup.sh`, navigate to a study that has human genes.  **If on staging**, Go to ["Human milk - differential expression"](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/human-milk-differential-expression#study-visualize) (SCP303).
2. Search for a gene
3. In related genes ideogram, hover over a gene
4. Confirm the tooltip shows RNA and protein diagrams
5. Play around with the new functionality, note anything that seems defective

Integrating this upstream Ideogram.js enhancement is a Developer's Choice Days project, tracked in SCP-4887.

### CODE REVIEW

**Review by bistline — APPROVED** [review]
Very cool - works as advertised from demo.  Adds a lot of rich information in a very compact UI.

My only nit is that tooltip on the button for "Unsplice exon (s)" takes a long time to load.  I know this is default behavior in all browsers for title tags on DOM elements, but it might be nicer if it could load faster somehow.  The reason I bring it up is that it is not immediately obvious what the button does - seeing it I would think that I was cutting content from somewhere, which didn't make sense.  Apologies if someone suggested this yesterday - I couldn't remember if it was brought up or not.

**Comment by eweitz** [comment]
> tooltip on the button for "Unsplice exon (s)" takes a long time to load. I know this is default behavior in all browsers for title tags on DOM elements, but it might be nicer if it could load faster somehow.

Good point.  I use `title` in a few other places in the broader gene tooltip, but I agree that bare icons warrant faster text display on hover.  In the next Ideogram.js version I'll plan to add a tiny dependency to do that, or simply write the likely trivial code for it.

**Review by michelle-vel — COMMENTED** [review]
Wow, the feature looks great! Looks great in the video

**Review by ehanna4 — APPROVED** [review]
Overall looks good and the hovering is nice and clean, fades nicely, the things are responsive, looks good! Only question I have is that Gene name at the top of the box appears to be a link but clicking it doesn't do anything. Is it supposed to do something when clicked? (see video attached)

https://user-images.githubusercontent.com/54322292/214084921-0057862a-f1b8-4874-afd8-f3d6342ea23f.mov

**Review by jlchang — COMMENTED** [review]
Very cool! (I haven't watched the video of last week's demo yet so apologies if any of my comments are redundant)

Can the default state of the RNA diagram be the pre-RNA state? I'm not sure whether the scissor icon would need to change... The pre-RNA diagram (that has introns, exons and UTRs) is a very familiar viz in biology and somewhat specific to mRNA visualizations. While the mRNA diagram (FYI, not technically "protein" - for a protein "diagram" I'd expect to see a protein structure, which is quite a different viz), wasn't immediately recognizable because colored rectangles are used for many other UI functionalities. 

It seems if I toggle the scissors in one gene window, it stays using the pre mRNA view for all the rest. Is that intentional?

I'm not able to see the protein features visualization in my local instance - is this user error on my part?

Also, in my functional testing, the MTOR gene popup didn't show a dropdown for alternative transcripts that is show in the video - is that expected?

Regarding alternative transcripts, for MTX files that use both columns in the genes file AND are 10x v3 5' experiments, the data may have distinct alternative transcript data so, technically, the portal user could request a particular RNA isoform... I don't know how many of our studies have such info... We should probably look to see how we report that data when multiple isoforms are associated with the same gene name.

**Comment by eweitz** [comment]
Thanks for the insightful scientific review, Jean.  

In our debugging session, we saw the local discrepancies were likely due to your instance running Ideogram.js 1.38.0 instead of 1.41.0.  Iterating with Docker on Intel Mac was very slow, and opted to do post-merge testing, especially given videos from Emily and me indicate the RNA and protein diagrams work as expected locally.

We also discussed pre-mRNA vs. mature mRNA as the default diagram.  The former is indeed more conventional.  We opted to begin by trying the latter as default, as it seems more useful since it gives a more coherent coordinate space to show aligned protein features.  We agreed the mature mRNA default would thus better help generate hypotheses, and so favored it.

Your notes on how some library preparation protocols capture preliminary mRNA were fantastic.  I hadn't appreciated that at all.  We concurred that something like library-aware display of pre- vs. mature-mRNA might be neat in the future, but isn't a blocker.

> Regarding alternative transcripts, for MTX files that use both columns in the genes file AND are 10x v3 5' experiments, the data may have distinct alternative transcript data so, technically, the portal user could request a particular RNA isoform... I don't know how many of our studies have such info... We should probably look to see how we report that data when multiple isoforms are associated with the same gene name.

This is a great point.  Your June 2022 analysis indicated the [prevalence of multiple transcripts per gene](https://docs.google.com/presentation/d/1SnOf3IS1lzdUgg-oKa8gp-Fc6QU_tV0M9U1aqj4RI2c/edit#slide=id.g13073d67193_0_118) is very low.  That said, perhaps there are a substantial number of studies that include transcript IDs, and only record 1 transcript, _and it is an alternative transcript_.  Analyzing questions like this indeed seems valuable.

In summary, per our chat, I'll go ahead and merge this now given the approvals, and we can easily revert if blockers are found when doing further scientific functional reviews on staging.
